src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/DynamicTable.cs

```diff
@@ -67,18 +69,15 @@ public void Insert(int? staticTableIndex, ReadOnlySpan<byte> name, ReadOnlySpan<
                 return;
             }
 
-            var entry = new HeaderField(staticTableIndex, name, value);
-            _buffer[_insertIndex] = entry;
-            _insertIndex = (_insertIndex + 1) % _buffer.Length;
-            _size += entry.Length;
-            _count++;
-        }
-
-        public void Resize(int maxSize)
-        {
-            if (maxSize > _maxSize)
+            // Ensure that we have at least one slot available.
+            if (_count == _buffer.Length)
             {
-                var newBuffer = new HeaderField[maxSize / HeaderField.RfcOverhead];
+                Debug.Assert(_count + 1 <= _maxSize / HeaderField.RfcOverhead);
+
+                int newBufferSize = Math.Min(Math.Max(4, _buffer.Length * 2), _maxSize / HeaderField.RfcOverhead);
```

Comment:
If someone is using the dynamic table feature, is it likely they'll only have 4 items in the table? Maybe I misunderstand typical use, but that seems small.

Reply:
It'll heavily depend on the use case, but you're right, 4 is likely to be exceeded.
E.g. if you sent a single request, this might be the number of response headers for that request.
If you're hitting an API where response headers look the same, it might never grow after that.
If you were to use HttpClient as a crawler, then you're practically guaranteed to either use 0 if the server doesn't use it, or maxSize if it does.